go/teams/box_audit.go

```diff
@@ -0,0 +1,1100 @@
+	return &l.Audits[len(l.Audits)-1]
+}
+
+// BoxAudit is a single sequence of audit attempts for a single team.
+type BoxAudit struct {
+	ID       BoxAuditID
+	Attempts []keybase1.BoxAuditAttempt
+}
+
+type BoxAuditID = []byte
+
+func NewBoxAuditID() (BoxAuditID, error) {
+	idBytes := make([]byte, BoxAuditIDLen)
+	_, err := rand.Read(idBytes)
+	if err != nil {
+		return nil, err
+	}
+	return BoxAuditID(idBytes), nil
+}
+
+// BoxAuditQueue holds a list of teams that need to be reaudited, because the
+// previously failed an audit. When a team does pass an audit, it is removed
+// from the queue.
+type BoxAuditQueue struct {
+	Items   []BoxAuditQueueItem
+	Version Version
+}
+
+var _ Versioned = &BoxAuditQueue{}
+
+func (q *BoxAuditQueue) GetVersion() Version {
+	return q.Version
+}
+
+func NewBoxAuditQueue(version Version) *BoxAuditQueue {
+	return &BoxAuditQueue{
+		Items:   nil,
+		Version: version,
+	}
+}
+
+type BoxAuditQueueItem struct {
+	Ctime      time.Time
+	TeamID     keybase1.TeamID
+	BoxAuditID BoxAuditID
+}
+
+// BoxAuditJail contains TeamIDs that have hit a fatal audit failure or the max
+// number of retryable audit failures. Teams in jail will not be reaudited
+// unless they are explicitly loaded by the fast or slow team loaders.
+type BoxAuditJail struct {
+	TeamIDs map[keybase1.TeamID]bool
+	Version Version
+}
+
+var _ Versioned = &BoxAuditJail{}
+
+func (j *BoxAuditJail) GetVersion() Version {
+	return j.Version
+}
+
+func NewBoxAuditJail(version Version) *BoxAuditJail {
+	return &BoxAuditJail{
+		TeamIDs: make(map[keybase1.TeamID]bool),
+		Version: version,
+	}
+}
+
+func (a *BoxAuditor) shouldAudit(mctx libkb.MetaContext, team Team) (bool, *keybase1.BoxAuditAttemptResult, error) {
+	if team.IsSubteam() {
+		res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_SUBTEAM
+		return false, &res, nil
+	}
+	if team.IsOpen() {
+		res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_OPENTEAM
+		return false, &res, nil
+	}
+	role, err := team.MemberRole(mctx.Ctx(), mctx.CurrentUserVersion())
+	if err != nil {
+		return false, nil, err
+	}
+	if !role.IsOrAbove(keybase1.TeamRole_WRITER) {
+		res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_ROLE
+		return false, &res, nil
+	}
+
+	return true, nil, nil
+}
+
+// loadTeamForBoxAudit loads a team once, but if the client
+// has not yet stored BoxSummaryHashes (due to being an old client)
+// it does a force full reload so it is populated.
+func loadTeamForBoxAudit(mctx libkb.MetaContext, teamID keybase1.TeamID) (*Team, error) {
+	return loadTeamForBoxAuditInner(mctx, teamID, false)
+}
+
+func loadTeamForBoxAuditInner(mctx libkb.MetaContext, teamID keybase1.TeamID, force bool) (team *Team, err error) {
+	defer mctx.TraceTimed("loadTeamForBoxAuditInner", func() error { return err })()
+	arg := keybase1.LoadTeamArg{
+		ID:              teamID,
+		ForceRepoll:     true,
+		Public:          teamID.IsPublic(),
+		ForceFullReload: force,
+	}
+
+	team, err = Load(mctx.Ctx(), mctx.G(), arg)
```

Comment:
Something I've been thinking about - maybe MetaContext in this function should have a different type to signify that it's passing `SkipBoxAuditCheckContextKey` flag. In general I like how `initMctx` works in this file, but it's a bit of work to trace back where `mctx` is created to ensure it has the right flags.

Reply:
interesting, like a wrapper struct around a mctx?

Reply:
hm I'm not sure. Maybe @maxtaco or @mlsteele have opinions on this. For this PR I think it's good as it is, but maybe as a further cleanup. I'm afraid that in future some other code path will call into this without the "right" context and things will break subtly

Reply:
ok, I generally think things like this are a good idea (preventing errors with types). I'm not sure exactly how it would work here though, let me know if you have an idea and I'll try to implement it